Panic when updating to v2.0.4-rc4

<!-- Please search for open issues that relate to the same problem before opening a new one. -->

### Describe the bug:
I updated to version v2.0.4-rc4 and tried to run my app but it panicked.
<!-- A clear and concise description about the bug. -->


### To Reproduce:
Steps to reproduce the behaviour:
1. Execute code
2. See error


### Example code:
<!-- If applicable, add a short code snippet to help explain and simplify reproduction of the problem. -->
```go
package main

import (
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/widget"
)

func main() {
	a := app.New()
	w := a.NewWindow("Test")
	w.SetContent(container.NewBorder(
		widget.NewForm(
			widget.NewFormItem("Nummer", widget.NewLabel("")),
		),
		nil, nil, nil, widget.NewEntry(),
	))
	w.ShowAndRun()
}
```

go.mod:
```
module github.com/Cookie04DE/FyneTest

go 1.16

require (
	fyne.io/fyne/v2 v2.0.4-rc4
)
```

Error:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x5c0666]

goroutine 27 [running, locked to thread]:
image.(*Uniform).RGBA(...)
        /media/conrad/Seagate/exclude_backup/Installationen/Go/src/image/names.go:29
image/draw.drawGlyphOver(0xc0000dc980, 0x1, 0x5, 0xb, 0x10, 0xc000098dc0, 0xc0003d2000, 0x0, 0x119c)
        /media/conrad/Seagate/exclude_backup/Installationen/Go/src/image/draw/draw.go:488 +0x86
image/draw.DrawMask(0x830898, 0xc0000dc980, 0x1, 0x5, 0xb, 0x10, 0x82fdc0, 0xc000098dc0, 0x0, 0x0, ...)
        /media/conrad/Seagate/exclude_backup/Installationen/Go/src/image/draw/draw.go:150 +0x1e7
golang.org/x/image/font.(*Drawer).DrawString(0xc0006955d0, 0x7c1e07, 0x6)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/golang.org/x/image@v0.0.0-20200430140353-33d19683fad8/font/font.go:178 +0x165
fyne.io/fyne/v2/internal/painter/gl.(*glPainter).newGlTextTexture(0xc0003f0150, 0x832cf8, 0xc0003ec4d0, 0xc0002aac50)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/painter/gl/gl_common.go:92 +0x348
fyne.io/fyne/v2/internal/painter/gl.getTexture(0x832cf8, 0xc0003ec4d0, 0xc000695750, 0xc0006956e8)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/painter/gl/gl_common.go:35 +0x94
fyne.io/fyne/v2/internal/painter/gl.(*glPainter).drawTextureWithDetails(0xc0003f0150, 0x832cf8, 0xc0003ec4d0, 0xc000695750, 0x4140000041400000, 0x41a8000042880000, 0x42ac000042fc0000, 0x0, 0x3f800000)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/painter/gl/draw.go:15 +0x45
fyne.io/fyne/v2/internal/painter/gl.(*glPainter).drawText(0xc0003f0150, 0xc0003ec4d0, 0x4140000041400000, 0x42ac000042fc0000)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/painter/gl/draw.go:82 +0x225
fyne.io/fyne/v2/internal/painter/gl.(*glPainter).drawObject(0xc0003f0150, 0x832cf8, 0xc0003ec4d0, 0x4140000041400000, 0x42ac000042fc0000)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/painter/gl/draw.go:101 +0x3da
fyne.io/fyne/v2/internal/painter/gl.(*glPainter).Paint(0xc0003f0150, 0x832cf8, 0xc0003ec4d0, 0x4140000041400000, 0x42ac000042fc0000)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/painter/gl/painter.go:70 +0x98
fyne.io/fyne/v2/internal/driver/glfw.(*glCanvas).paint.func1(0xc0000dc8c0, 0x4140000041400000)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/canvas.go:439 +0xe3
fyne.io/fyne/v2/internal/driver/glfw.(*glCanvas).walkTree.func1(0x832cf8, 0xc0003ec4d0, 0x4140000041400000, 0x0, 0x4f0000004f000000, 0x699600)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/canvas.go:535 +0x142
fyne.io/fyne/v2/internal/driver.walkObjectTree(0x832cf8, 0xc0003ec4d0, 0x4080000040800000, 0x8329f8, 0xc0003d2180, 0x4080000040800000, 0x0, 0x4f0000004f000000, 0xc000695d28, 0xc000695d00, ...)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:169 +0x482
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:176
fyne.io/fyne/v2/internal/driver.walkObjectTree(0x8329f8, 0xc0003d2180, 0x4080000040800000, 0x8329f8, 0xc0003d2240, 0x4080000040800000, 0x0, 0x4f0000004f000000, 0xc000695d28, 0xc000695d00, ...)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:190 +0x3ff
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:176
fyne.io/fyne/v2/internal/driver.walkObjectTree(0x8329f8, 0xc0003d2240, 0x4080000040800000, 0x833238, 0xc0000e0420, 0x4080000040800000, 0x0, 0x4f0000004f000000, 0xc000695d28, 0xc000695d00, ...)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:190 +0x3ff
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:176
fyne.io/fyne/v2/internal/driver.walkObjectTree(0x833238, 0xc0000e0420, 0x4080000040800000, 0x8329f8, 0xc0000dc4c0, 0x4080000040800000, 0x0, 0x4f0000004f000000, 0xc000695d28, 0xc000695d00, ...)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:190 +0x3ff
fyne.io/fyne/v2/internal/driver.walkObjectTree.func1(...)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:176
fyne.io/fyne/v2/internal/driver.walkObjectTree(0x8329f8, 0xc0000dc4c0, 0x8b94c100, 0x0, 0x0, 0x0, 0x0, 0x4f0000004f000000, 0xc000695d28, 0xc000695d00, ...)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:190 +0x3ff
fyne.io/fyne/v2/internal/driver.WalkVisibleObjectTree(0x8329f8, 0xc0000dc4c0, 0xc000695d28, 0xc000695d00, 0x48d625)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/util.go:134 +0x94
fyne.io/fyne/v2/internal/driver/glfw.(*glCanvas).walkTree(0xc0002be000, 0xc0000ca020, 0xc000695e00, 0xc000695de8)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/canvas.go:556 +0x15c
fyne.io/fyne/v2/internal/driver/glfw.(*glCanvas).walkTrees(0xc0002be000, 0xc000695e00, 0xc000695de8)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/canvas.go:563 +0x4f
fyne.io/fyne/v2/internal/driver/glfw.(*glCanvas).paint(0xc0002be000, 0x42ac000042fc0000)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/canvas.go:453 +0x129
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).repaintWindow.func1()
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/loop.go:164 +0x88
fyne.io/fyne/v2/internal/driver/glfw.(*window).RunWithContext(0xc00010c1c0, 0xc000052ea8)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/window.go:1207 +0x4f
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).repaintWindow(0xc0000c2370, 0xc00010c1c0)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/loop.go:155 +0x65
fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).startDrawThread.func1(0xc0000c2370, 0xc00008e660, 0xc0000c24b0)
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/loop.go:211 +0x309
created by fyne.io/fyne/v2/internal/driver/glfw.(*gLDriver).startDrawThread
        /media/conrad/Seagate/Programmieren/Go/src/pkg/mod/fyne.io/fyne/v2@v2.0.4-rc4/internal/driver/glfw/loop.go:182 +0xbd
exit status 2
```


### Device (please complete the following information):
 - **OS:** Linux Ubuntu<!-- [e.g. Linux, MacOS or iOS] -->
 - **Version:** Ubuntu 20.04.2 LTS<!-- [e.g. 5.10.2, 10.13 High Sierra or 14.2] -->
 - **Go version:** go1.16.6 linux/amd64<!-- [e.g. 1.12.3] -->
 - **Fyne version:** v2.0.4-rc4<!-- [e.g. 2.0.3 or git SHA] -->

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,7 +3,7 @@
 This file lists the main changes with each version of the Fyne toolkit.
 More detailed release notes can be found on the [releases page](https://github.com/fyne-io/fyne/releases). 
 
-## 2.0.4 - 4 August 2021
+## 2.0.4 - 6 August 2021
 
 ### Changed
 
@@ -28,6 +28,8 @@ More detailed release notes can be found on the [releases page](https://github.c
 * ScrollToBottom not always scrolling all the way when items added to container.Scroller
 * Fixed scrollbar disappearing after changing content (#2303)
 * Calling SetContent a second time with the same content will not show
+* Drawing text can panic when Color is nil (#2347)
+* Optimisations when drawing transparent rectangle or whitespace strings
 
 
 ## 2.0.3 - 30 April 2021

--- a/internal/painter/gl/gl_common.go
+++ b/internal/painter/gl/gl_common.go
@@ -5,6 +5,7 @@ import (
 	"log"
 	"runtime"
 
+	"fyne.io/fyne/v2/theme"
 	"github.com/goki/freetype"
 	"github.com/goki/freetype/truetype"
 	"golang.org/x/image/font"
@@ -72,6 +73,10 @@ func (p *glPainter) newGlStrokedRectTexture(obj fyne.CanvasObject) Texture {
 
 func (p *glPainter) newGlTextTexture(obj fyne.CanvasObject) Texture {
 	text := obj.(*canvas.Text)
+	color := text.Color
+	if color == nil {
+		color = theme.ForegroundColor()
+	}
 
 	bounds := text.MinSize()
 	width := int(p.textureScale(bounds.Width))
@@ -86,7 +91,7 @@ func (p *glPainter) newGlTextTexture(obj fyne.CanvasObject) Texture {
 
 	d := font.Drawer{}
 	d.Dst = img
-	d.Src = &image.Uniform{C: text.Color}
+	d.Src = &image.Uniform{C: color}
 	d.Face = face
 	d.Dot = freetype.Pt(0, height-face.Metrics().Descent.Ceil())
 	d.DrawString(text.Text)

--- a/internal/painter/software/draw.go
+++ b/internal/painter/software/draw.go
@@ -9,6 +9,7 @@ import (
 	"fyne.io/fyne/v2/canvas"
 	"fyne.io/fyne/v2/internal"
 	"fyne.io/fyne/v2/internal/painter"
+	"fyne.io/fyne/v2/theme"
 
 	"github.com/goki/freetype"
 	"github.com/goki/freetype/truetype"
@@ -139,6 +140,11 @@ func drawText(c fyne.Canvas, text *canvas.Text, pos fyne.Position, base *image.N
 	height := internal.ScaleInt(c, bounds.Height)
 	txtImg := image.NewRGBA(image.Rect(0, 0, width, height))
 
+	color := text.Color
+	if color == nil {
+		color = theme.ForegroundColor()
+	}
+
 	var opts truetype.Options
 	fontSize := text.TextSize * c.Scale()
 	opts.Size = float64(fontSize)
@@ -147,7 +153,7 @@ func drawText(c fyne.Canvas, text *canvas.Text, pos fyne.Position, base *image.N
 
 	d := font.Drawer{}
 	d.Dst = txtImg
-	d.Src = &image.Uniform{C: text.Color}
+	d.Src = &image.Uniform{C: color}
 	d.Face = face
 	d.Dot = freetype.Pt(0, height-face.Metrics().Descent.Ceil())
 	d.DrawString(text.Text)

--- a/widget/form.go
+++ b/widget/form.go
@@ -111,6 +111,7 @@ func (f *Form) createInput(item *FormItem) fyne.CanvasObject {
 func (f *Form) createLabel(text string) *canvas.Text {
 	return &canvas.Text{Text: text,
 		Alignment: fyne.TextAlignTrailing,
+		Color:     theme.ForegroundColor(),
 		TextSize:  theme.TextSize(),
 		TextStyle: fyne.TextStyle{Bold: true}}
 }

